### PR TITLE
implement Helium NBT

### PR DESCRIPTION
**Pull request by ExaInsanity**
This pull request aims to implement a fully-functional NBT library into Helium.

Adds: NBT, NBT docs, some packets and packet data

@Zefyro should we host the docs into github pages?

### CODE REVIEW

File: src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+
+using Helium.Nbt.Exceptions;
+using Helium.Nbt.Internal;
+
+[RequiresPreviewFeatures]
+public class RecursiveBinaryNbtReader
+{
+	public NbtTokenType CurrentToken { get; set; }
```
Comment (Zefyro) [inline src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs:16]: Wouldn't it be better to call this `CurrentNbtToken`?

File: src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+		if(type > (Byte)NbtTokenType.LongArray)
+		{
+			throw new MalformedNbtException($"Invalid token type found: {type}");
+		}
+
+		return (NbtTokenType)type;
+	}
+
+	public INbtToken ReadCurrentToken(NbtTokenType tokenType, Boolean readName = true)
```
Comment (Zefyro) [inline src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs:37]: Same here `ReadCurrentNbtToken`

**Review by Zefyro — COMMENTED** [review]
Other than those 2 I'm fine with these, and even if you don't change those it doesn't matter that much.

**Comment by Zefyro** [comment]
nvm, that was all in `project-helium/pull/9` well just think about the stuff in `src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs` the other part of the review isn't done yet

**Comment by ExaInsanity** [comment]
The question was more about the parsing approach (recursion vs flat loop) but thanks for the feedback, i suppose 😛

**Comment by ExaInsanity** [comment]
i'll just take this as an implicit "yes, go for it" then

**Comment by Zefyro** [comment]
well I didn't have anything to complain about the approach, so yes, take it as "yes, go for it".

File: src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs (hunk 1 above)
Comment (ExaInsanity) [reply src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs:16]: i dont think theres any need to specify that, the code is already verbose as is; and its `Helium.NBT.Serialization.BinaryNbtReader` already :p

Comment (Zefyro) [reply src/Helium.Nbt/Serialization/RecursiveBinaryNbtReader.cs:16]: I guess... :P

**Comment by ExaInsanity** [comment]
image
@Zefyro benchmarked the different readers, i propose all of Helium uses RecursiveBigEndian internally

**Comment by ExaInsanity** [comment]
PR effectively finished, awaiting review @Zefyro

**Review by Zefyro — APPROVED** [review]
Looking alright. I don't see anything that would kill me slowly and painfully.

**Comment by Zefyro** [comment]
Do I have to... thats scary...

**Comment by ExaInsanity** [comment]
:gun:

**Pull request merged**